### PR TITLE
logger update

### PR DESCRIPTION
**Pull request by burner**
this is a breaking change.
In order to make the logger work with allocator and the forward messages
without the need to gc allocation writeLogMsg had to go.
To test the design a new logger was added.
This logger is called AllocLogger and has an additional ctor argument,
an Allocator.

Additionally, the unit tests were simplified.

### CODE REVIEW

**Comment by burner** [comment]
not sure why the auto-tester is failing, problem seems to be in std.algorithm

**Comment by wilzbach** [comment]
Looks like a spurious failure.

File: std/experimental/logger/alloclogger.d [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            import std.string : lastIndexOf;
+            ptrdiff_t fnIdx = file.lastIndexOf('/') + 1;
+            ptrdiff_t funIdx = funcName.lastIndexOf('.') + 1;
+
+            ()@trusted
+            {
+                systimeToISOString(this.oRange, timestamp);
+                formattedWrite(this.oRange, ":%s:%s:%u ", file[fnIdx .. $],
+                    funcName[funIdx .. $], line);
```
[the remark this answers is not in the document]
Comment (burner) [reply std/experimental/logger/alloclogger.d]: true, but this PR is at an impasse. if the three abstract method in Logger are @safe most loggers can be @safe which allows them to be used in @safe code. AllocLogger might not be actually @safe. If the three methods are @system you can not use logger in @safe code. I'm not sure what to do, but I'm open to suggestions.

File: std/experimental/logger/alloclogger.d [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+module std.experimental.logger.alloclogger;
+
+import std.experimental.logger.core;
+
+private struct AllocOutRange(Alloc)
+{
+    /// The used allocator
+    Alloc* alloc;
+    /// the output
+    char[]* str;
```
Comment (wilzbach) [inline std/experimental/logger/alloclogger.d]: An array is already a pointer? Why does it need to be a pointer to a pointer?

File: std/experimental/logger/alloclogger.d (hunk 1 above)
[the remark this answers is not in the document]
Comment (wilzbach) [reply std/experimental/logger/alloclogger.d]: Better make it @system for now than lying about safety?
A future version could do some smart `static if` to do the @trusting for cases when it's known that it's absolutely @safe?

**Comment by burner** [comment]
This is a classic catch-22 situation. The abstract class Logger has three abstract methods that are @safe to make logging usable in @safe code. Either I make them @system and thereby disable the use of logger in @safe code or I mark the AllocLogger code as @trusted, to keep the logger usable in @safe code. Or I delete the AllocLogger. I'm open to suggestions.

**Comment by LightBender** [comment]
PR closed as stalled. This can be resurrected for Phobos 3.

**Pull request closed without merging**